### ISSUE TITLE
Improve diffing of values: Concrete Types

### ISSUE THREAD

**Issue by Siddhi-agg** [NathanReb/ocaml-api-watch#45]
In our ongoing effort to improve the accuracy and reliability of our interface diffing tool, we need to ensure that our implementation can correctly identify modifications to values that reference **concrete types (records or variants)**. At present, our tool may fail to detect certain modifications or incorrectly flag identical concrete types as different. 

For example:
```
type t = { a: int; b: string }
val x : t
```
If we pass this same interface given above form two different cmi, they should be treated as identical and no modification should be detected in `val x`. 

Only when we have modified the record defined in `type t` like in the interface given below, should a change in `val f` be detected.
```
type t = { a: int; b: float }
val x : t
```

Similarly for variants of a type, a change in a value which references the type should only be detected when atleast one of the variants for the type has changed.

This issue will address these shortcomings by adding relevant test cases and updating the implementation accordingly to make the added tests pass.

**Comment by NathanReb** [comment]
> Only when we have modified the record defined in type t like in the interface given below, should a change in val f be detected.

I think in this specific case, that might not be true. We decided to mark the change if the declared type is just an alias because there are other legitimate ways to build values of those type, e.g. if somewhere one defines an exposed alias such as `type t = int list` and provide values and functions referencing the type `t`, they could be legitimately tempted to use it on arbitrary lists of integers.

Here, given the type is defined in this module, I don't think we should mark values referencing it as modified (unless they've been otherwise modified) since the type does not exist outside of this module, as opposed to lists of integers for instance.

I think in that case, an external user of the library defining `t` that updates to this new version that modifies the definition of `t` would have to fix pattern matchings over values of type `t` and "literals" of type `t` and that would be enough of a fix, they would not have to fix a call to a `val f : t -> string` function for example. What I'm trying to say is that in this case, from a library user perspective, `x` or the aforementioned `f` would not have changed.

Am I making sense here? I'm finding this harder to describe than I'd hoped.

**Comment by NathanReb** [comment]
Also to clarify, I think the expected semantic would be that considering the following two interfaces:
```ocaml
type t = {a : int; b : string}
val x : t
```
and
```ocaml
type t = {a : int; b : float}
val x : t
```
we should report a change in `type t` but no change for `val x`.

**Comment by Siddhi-agg** [comment]
Okay @NathanReb, I think I understand what you mean! I see your point regarding the distinction between type aliases and defined types within the module. Thanks for providing such an insightful explanation!

I had a follow up question: If we want to detect just the changes in the type definitions in these cases mentioned above, that would be dealt when we are writing implementation for detecting type changes right? For now, we will just focus on not detecting the value changes from a variant or record change, am I correct?

**Comment by NathanReb** [comment]
Yes exactly, the change in the type will simply by marked as `[Any]` and indeed we're testing that no value change is reported!
